### PR TITLE
back-porting DeepTauId v2 from branch cms-tau-pog:CMSSW_10_6_X_tau-pog_DeepTau2017v2

### PR DESCRIPTION
**Pull request by MRD2F**
Changes regarding back-porting DeepTauId v2 integration from the PR #123 from 106X to 102X

Studies of the measure memory consumption and cpu timing have been preformed, and the files with the complete information are attached, both tests have been preformed for 1000 events with a 2017
TTToHadronic sample

The total memory consumption is 21.5 MiB, calculated as the sum of the three bigger contributions:

1. DeepTauId::getPredictions =  17.2 MiB
2. deep_tau::DeepTauBase::initializeGlobalCache = 3.9 MiB
3. DeepTauId::DeepTauId = 0.4 MiB

[igprof-analyse-DeepTau_102X.txt.gz](https://github.com/cms-tau-pog/cmssw/files/3188363/igprof-analyse-DeepTau_102X.txt.gz)

Together (DeepTauId after deepJet) the memory consumption by DeepTauId has decrease. The total memory consumption with deepJet is 6.4 MiB, calculated also as the sum of the three bigger contributions:

1. DeepTauId::getPredictions =  3.22 MiB
2. deep_tau::DeepTauBase::initializeGlobalCache = 2.73 MiB
3. DeepTauId::DeepTauId = 0.41 MiB

[igprof-analyse-DeepTau_102X_deepJet.txt.gz](https://github.com/cms-tau-pog/cmssw/files/3188367/igprof-analyse-DeepTau_102X_deepJet.txt.gz)

The total cpu timing is 279.45 s. The DeepTau::Produce function takes 250.74 seconds to process 1000 events, so has a frequency of 3.99 Hz.
[DeepTauOriginalCPU_102X.txt.gz](https://github.com/cms-tau-pog/cmssw/files/3188369/DeepTauOriginalCPU_102X.txt.gz)

### CODE REVIEW

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 1]
```diff
@@ -291,32 +634,695 @@ class DeepTauId : public deep_tau::DeepTauBase {
...
+              { "input_inner_muon", *muonInnerTensor_ }, { "input_outer_muon", *muonOuterTensor_ },
+              { "input_inner_hadrons", *hadronsInnerTensor_ }, { "input_outer_hadrons", *hadronsOuterTensor_ } },
+            { "main_output/Softmax" }, &pred_vector);
+
+    }
+
+    template<typename Collection>
+    void fillGrids(const TauType& tau, const Collection& objects, CellGrid& inner_grid, CellGrid& outer_grid)
+    {
+        static constexpr double outer_dR2 = std::pow(0.5, 2);
```
Comment (mbluj) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: Do not initialize `static constexpr` with a function but by a number. Fixed in the 10_6_X version by commit 8fcee1d3599daa2d0ba310dcb782f986d0210237

File: RecoTauTag/RecoTau/test/runDeepTauIDsOnMiniAOD.py [hunk 2]
```diff
@@ -77,3 +78,5 @@
      numberOfThreads = cms.untracked.uint32(nThreads),
      numberOfStreams = cms.untracked.uint32(0)
 )
+from Validation.Performance.IgProfInfo import customise
```
Comment (mbluj) [inline RecoTauTag/RecoTau/test/runDeepTauIDsOnMiniAOD.py]: I think IGProf is not needed for a regular user - please remove it.

**Review by mbluj — COMMENTED** [review]
A few issues to fix as commented inline.

**Comment by mbluj** [comment]
Please also include the d46fc5dc8dbcb45c63c0f29023d0f2a250552c32 and dfa3db3bcf4e2974894cb06ccfd0a6ebdcd2343e commits which solve ambiguity in definition of `MuonSubdetId`.

**Comment by kandrosov** [comment]
> Please also include the [d46fc5d](https://github.com/cms-tau-pog/cmssw/commit/d46fc5dc8dbcb45c63c0f29023d0f2a250552c32) and [dfa3db3](https://github.com/cms-tau-pog/cmssw/commit/dfa3db3bcf4e2974894cb06ccfd0a6ebdcd2343e) commits which solve ambiguity in definition of `MuonSubdetId`.

I have some doubts if [d46fc5d](https://github.com/cms-tau-pog/cmssw/commit/d46fc5dc8dbcb45c63c0f29023d0f2a250552c32) should be included this PR, because it affects ```DataFormats/MuonDetId``` package. As a consequence, ```git cms-merge-topic``` on top of CMSSW_10_2_14 would require to recompile significant about of packages that would make it difficult to use deepTau before it is integrated into the official CMSSW release (which could take time). Perhaps, it is better to have two separate branches: one which that could be provided for the analysists as soon as this PR is merged and other for the back-port integration into the central CMSSW. What do you think?

**Comment by mbluj** [comment]
It makes sense. Please include earlier comments and then I will clone the branch. Hopefully there will not be another changes. One can also check `unsafe` merge which can be simpler and in this case it should be safe.

**Comment by mbluj** [comment]
I have just checked with 10_6_X branch that the following works fine.
```
git cms-merge-topic -u cms-tau-pog:CMSSW_10_6_X_tau-pog_DeepTau2017v2
```
Therefore, I think it is better to not multiply branches, i.e. have unique 10_2_X branch with all fixes for both users and integration with CMSSW.

**Comment by mbluj** [comment]
Add also this one 9bdaa99a4fa33012653644b04d7548ffcc5852da
and this 38c6ff10217019d9c1438ee74b6c65f833a2505e fixing bug introduced in previous one.

**Comment by mbluj** [comment]
It is a backport to 102X from 106X identical to original modulo adaptation to differences between the two release series and lack of inclusion to the regular MiniAOD workflow. Successfully tested with provided test workflow.
=> Merging
=> Possible changes which can appear in process of integration with official CMSSW will be directly pushed to tau-pog repository.

**Pull request merged**